File: brainfuck/bf.go
package main

import (
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"runtime"
	"log"
)

const (
	INC = iota
	MOVE
	PRINT
	LOOP
)

type Op struct {
	O    int
	V    int
	Loop []Op
}

func NewOp(op int, v int) Op {
	return Op{O: op, V: v}
}

func NewOpLoop(op int, l []Op) Op {
	return Op{O: op, Loop: l}
}

type StringIterator struct {
	Text []byte
	Pos  int
}

func NewStringIterator(s string) *StringIterator {
	si := &StringIterator{Text: []byte(s), Pos: 0}
	return si
}

func (si *StringIterator) Next() byte {
	if si.Pos < len(si.Text) {
		res := si.Text[si.Pos]
		si.Pos += 1
		return res
	} else {
		return byte(0)
	}
}

type Tape struct {
	tape []int
	pos  int
}

func NewTape() *Tape {
	t := &Tape{pos: 0}
	t.tape = make([]int, 1)
	return t
}

func (t *Tape) Inc(x int) {
	t.tape[t.pos] += x
}

func (t *Tape) Move(x int) {
	t.pos += x
	for t.pos >= len(t.tape) {
		t.tape = append(t.tape, 0)
	}
}

func (t *Tape) Get() int {
	return t.tape[t.pos]
}

type Printer struct {
	sum1 int
	sum2 int
	quiet bool
}

func NewPrinter(quiet bool) *Printer {
	return &Printer{sum1: 0, sum2: 0, quiet: quiet}
}

func (p *Printer) Print(n int) {
	if p.quiet {
		p.sum1 = (p.sum1 + n) % 255
		p.sum2 = (p.sum2 + p.sum1) % 255
	} else {
		fmt.Printf("%c", n)
	}
}

func (p *Printer) GetChecksum() int {
	return (p.sum2 << 8) | p.sum1
}

type Program struct {
	Ops []Op
}

func NewProgram(code string) *Program {
	return &Program{Ops: parse(NewStringIterator(code))}
}

func (p *Program) Run(printer *Printer) {
	_run(p.Ops, NewTape(), printer)
}

func parse(si *StringIterator) []Op {
	res := make([]Op, 0)
	for true {
		c := si.Next()
		var op Op
		switch c {
		case '+':
			op = NewOp(INC, 1)
		case '-':
			op = NewOp(INC, -1)
		case '>':
			op = NewOp(MOVE, 1)
		case '<':
			op = NewOp(MOVE, -1)
		case '.':
			op = NewOp(PRINT, 0)
		case '[':
			op = NewOpLoop(LOOP, parse(si))
		case ']':
			return res
		case byte(0):
			return res
		default:
			continue
		}
		res = append(res, op)
	}
	return res
}

func _run(program []Op, tape *Tape, p *Printer) {
	for i := 0; i < len(program); i++ {
		switch program[i].O {
		case INC:
			tape.Inc(program[i].V)
		case MOVE:
			tape.Move(program[i].V)
		case LOOP:
			for tape.Get() > 0 {
				_run(program[i].Loop, tape, p)
			}
		case PRINT:
			p.Print(tape.Get())
		}
	}
}

func notify(msg string) {
	conn, err := net.Dial("tcp", "localhost:9001")
	if err == nil {
		fmt.Fprintf(conn, msg)
		conn.Close()
	}
}

func verify() {
	text := `++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>
---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.`
	p_left := NewPrinter(true)
	NewProgram(text).Run(p_left)
	left := p_left.GetChecksum()

	p_right := NewPrinter(true)
	for _, c := range "Hello World!\n" {
		p_right.Print(int(c))
	}
	right := p_right.GetChecksum()
	if left != right {
		log.Fatalf("%+v != %+v\n", left, right)
	}
}

func main() {
	verify()
	code, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		panic(fmt.Sprintf("%v", err))
	}
	text := string(code)
	p := NewPrinter(os.Getenv("QUIET") != "")

	notify(fmt.Sprintf("%s\t%d", runtime.Compiler, os.Getpid()))
	NewProgram(text).Run(p)
	notify("stop")

	if p.quiet {
		fmt.Printf("Output checksum: %d\n", p.GetChecksum())
	}
}

File: brainfuck/bf.nim
import net
import os
import posix
import strformat

type
  OpType = enum Inc, Move, Loop, Print
  Ops = seq[Op]
  Op = object
    case op: OpType
    of Inc, Move: val: int
    of Loop: loop: Ops
    else: discard

  StringIterator = iterator(): char
  Tape = object
    pos: int
    tape: seq[int]

  Printer = object
    sum1: int
    sum2: int
    quiet: bool

  Program = distinct Ops

proc newTape(): Tape =
  Tape(pos: 0, tape: newSeq[int](1))

proc get(t: Tape): int =
  t.tape[t.pos]

proc inc(t: var Tape, x: int) =
  t.tape[t.pos] += x

proc move(t: var Tape, x: int) =
  t.pos += x
  while t.pos >= t.tape.len:
    t.tape.setLen 2 * t.tape.len

func newStringIterator(s: string): StringIterator =
  result = iterator(): char =
             for i in s:
               yield i

proc newPrinter(quiet: bool): Printer =
  Printer(sum1: 0, sum2: 0, quiet: quiet)

proc print(p: var Printer, n: int) =
  if p.quiet:
    p.sum1 = (p.sum1 + n) mod 255
    p.sum2 = (p.sum2 + p.sum1) mod 255
  else:
    stdout.write n.chr()
    stdout.flushFile()

proc checksum(p: Printer): int =
  (p.sum2 shl 8) or p.sum1

proc parse(iter: StringIterator): Ops =
  for i in iter():
    case i
    of '+': result.add Op(op: Inc, val: 1)
    of '-': result.add Op(op: Inc, val: -1)
    of '>': result.add Op(op: Move, val: 1)
    of '<': result.add Op(op: Move, val: -1)
    of '.': result.add Op(op: Print)
    of '[': result.add Op(op: Loop, loop: parse iter)
    of ']': break
    else: discard

proc parse(code: string): Program =
  let iter = newStringIterator(code)
  result = Program parse iter

proc run(ops: Ops, t: var Tape, p: var Printer) =
  for op in ops:
    case op.op
    of Inc: t.inc op.val
    of Move: t.move op.val
    of Loop:
      while t.get() > 0: run(op.loop, t, p)
    of Print:
      p.print(t.get())

proc run(ops: Program, p: var Printer) =
  var tape = newTape()
  run Ops ops, tape, p

proc notify(msg: string) =
  try:
    var socket = newSocket()
    defer: socket.close()
    socket.connect("localhost", Port(9001))
    socket.send(msg)
  except:
    discard

proc verify() =
    let text = """++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>
    ---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++."""
    var p_left = newPrinter(true)
    text.parse().run(p_left)
    let left = p_left.checksum

    var p_right = newPrinter(true)
    for c in "Hello World!\n":
        p_right.print(ord(c))
    let right = p_right.checksum
    if left != right:
      stderr.writeLine(&"{left} != {right}")
      quit(1)

when isMainModule:
  verify()
  let text = paramStr(1).readFile()
  var p = newPrinter(existsEnv("QUIET"))

  var compiler = "Nim/clang"
  when defined(gcc):
    compiler = "Nim/gcc"
  notify(&"{compiler}\t{getpid()}")
  text.parse().run(p)
  notify("stop")

  if p.quiet:
    echo &"Output checksum: {p.checksum}"
